src/backend_model_instance.cc

```diff
@@ -220,8 +220,7 @@ TritonModelInstance::SetInstances(
     }
     for (int32_t c = 0; c < group.count(); ++c) {
       std::string instance_name{
-          group.count() > 1 ? group.name() + "_" + std::to_string(c)
-                            : group.name()};
+          c > 0 ? group.name() + "_" + std::to_string(c) : group.name()};
```

Comment:
Nice detailed explanation. Two comments:

> In other words, all group_0 will be named group.

1. I was thinking the opposite, we should change "group" to "group_0" to have a consistent naming scheme regardless of count. 

2. What do you think about including device type / kind in the name? ex: "bert_cpu0_instance0", "bert_cpu0_instance1", "bert_gpu0_instance0", etc.

I think this would improve the intuition in the logging and debugging from user perspective. For example it may be more clear that a GPU instance is receiving work rather than a CPU instance in logs, compared to current naming convention.

Reply:
> I was thinking the opposite, we should change "group" to "group_0" to have a consistent naming scheme regardless of count.

I am good with any scheme, as long as the naming is consistent. Update commit: [Use different instance name unification method](https://github.com/triton-inference-server/core/pull/231/commits/af88121df20fa8d578e1d7a6a791b8f72a203a78)

Reply:
> What do you think about including device type / kind in the name? ex: "bert_cpu0_instance0", "bert_cpu0_instance1", "bert_gpu0_instance0", etc.

This is a good idea. I think users generally do not supply an instance name if it is optional, but when debugging is needed, they would want to see descriptive names.

But I think this change is out of scope for this PR, because
- [The code](https://github.com/triton-inference-server/core/blob/main/src/model_config_utils.cc#L779-L788) where the default name is assigned to instances is on a different file, independent from this change; and
- We can have more time design on the default instance name for different scenarios.

What do you think about the scope cutoff? On this PR, we only address how to assign different names to instances on the same instance group (i.e. `"group_default_name_0", "group_default_name_1", ...`). We can decide on the `group_default_name` on a different PR.

Reply:
Looking at how the default group name is set and the signature check, so if user change the config from:
```
instance_group [
  { kind: KIND_CPU },
  { kind: KIND_GPU },
]
```
to
```
instance_group [
  { kind: KIND_GPU },
  { kind: KIND_CPU },
]
```
No instance will be reused?

Reply:
Correct, because of the default name assigned to the instances will change if the location of the instances on the configuration changes.

Reply:
This is the reason why I replaced the [test_gpu_cpu_instance_update](https://github.com/triton-inference-server/server/pull/6018/files#diff-40ca2af021f4288f3fc6ae6071a244bca6350a0e657ab2822b61650cac266931L190) test with the [test_instance_name_update](https://github.com/triton-inference-server/server/pull/6018/files#diff-40ca2af021f4288f3fc6ae6071a244bca6350a0e657ab2822b61650cac266931R190) and [test_instance_same_name](https://github.com/triton-inference-server/server/pull/6018/files#diff-40ca2af021f4288f3fc6ae6071a244bca6350a0e657ab2822b61650cac266931R211) tests.

Reply:
Thanks for making the change for default name to be consistent with `group_0` regardless of count.

Regarding this:
> But I think this change is out of scope for this PR, because

>
> [The code](https://github.com/triton-inference-server/core/blob/main/src/model_config_utils.cc#L779-L788) where the default name is assigned to instances is on a different file, independent from this change; and
We can have more time design on the default instance name for different scenarios.
What do you think about the scope cutoff? On this PR, we only address how to assign different names to instances on the same instance group (i.e. "group_default_name_0", "group_default_name_1", ...). We can decide on the group_default_name on a different PR.

I'm fine with cutting off scope to separate PR. Please make a ticket and link it.

Reply:
This behavior is counter-intuitive, what is the reason of doing so?

Reply:
It is because of the logic which assigns default name to instances that do not specify a name.

I think there is a way to avoid the behavior, but it will be complicated: During an instance update, leave the default instance name empty, so the lookup can be done purely on the instance config. Instances with the name specified will be matched first, and then for those that do not specify a name. Those without a name will use the name on the instance that is matched, if any. Otherwise, a default name is assigned.

Maybe, we can simply print on the log notifying the user when an instance reused has a different name than the one requested.

Reply:
Other than minimizing the scope of the code change. I agree the behavior is counter-intuitive. Maybe a redesign on how the default name works is needed.